### PR TITLE
[COST-5544] Make leader election duration options configurable

### PR DESCRIPTION
**Pull request by djnakabaale**

### CODE REVIEW

**Review by maskarb — COMMENTED** [review]
This is looking good! Tho I'm not sure we need those get-env functions in a separate file. I would just put them in `main.go` as internal functions for the time being and then move them out if there is a plan to use them elsewhere.

File: internal/utils/misc_utils.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+			logger.Error(err, "Invalid boolean format for environment variable", "Variable", varName, "Value", value)
+		}
+		return parsedVal
+	}
+	return defaultValue
+}
+
+// GetEnvVar returns the value from an environment variable or the
+// provided default value if the variable does not exist.
+func GetEnvVar(varName, defaultValue string) string {
```
[the remark this answers is not in the document]
Comment (djnakabaale) [reply internal/utils/misc_utils.go]: I've updated the function names to be more descriptive.

File: internal/utils/misc_utils.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		return defaultValue
+	}
+
+	if parsedVal, err := time.ParseDuration(val); err == nil {
+		return parsedVal
+
+	} else {
+		logger.Error(err, "Invalid boolean format for environment variable", "Variable", varName, "Value", val)
+		return defaultValue
+	}
```
[the remark this answers is not in the document]
Comment (djnakabaale) [reply internal/utils/misc_utils.go]: Updated this function to take a valid time.Duration value as default.

File: internal/utils/misc_utils.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// provided default boolean value if the variable does not exist.
+func GetEnvVarBool(varName string, defaultValue bool) bool {
+	if value, exists := os.LookupEnv(varName); exists {
+		parsedVal, err := strconv.ParseBool(value)
+		if err != nil {
+			logger.Error(err, "Invalid boolean format for environment variable", "Variable", varName, "Value", value)
+		}
+		return parsedVal
+	}
+	return defaultValue
```
[the remark this answers is not in the document]
Comment (djnakabaale) [reply internal/utils/misc_utils.go]: This is now updated and more readable.

File: cmd/main.go [hunk 4]
```diff
@@ -48,14 +50,23 @@ func init() {
...
 	var probeAddr string
+	var enableLeaderElection bool
 	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
 	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
 	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
 
+	// fetch leader election configurations from environment variables
+	enableLeaderElection = utils.GetEnvVarBool("LEADER_ELECTION_ENABLED", false)
```
[the remark this answers is not in the document]
Comment (djnakabaale) [reply cmd/main.go]: Nice catch, this was left over in [this commit](https://github.com/project-koku/koku-metrics-operator/pull/459/commits/62e88d4c655b2a9c5c407d7788d30c9cff9a8602), from experimenting with turning on/off leader election. It's updated now.

**Review by maskarb — APPROVED** [review]
This looks good to me. We can skip patch coverage for the rest of this PR. Thanks for making those other changes.

File: cmd/main.go [hunk 5]
```diff
@@ -138,3 +150,50 @@ func getWatchNamespace() (string, error) {
...
+	return parsedVal
+
+}
+
+// validateLeaderElectionConfig returns the Namespace the operator should be watching for changes
+func validateLeaderElectionConfig(leaseDuration, renewDeadline, retryPeriod time.Duration) (time.Duration, time.Duration, time.Duration) {
+
+	// validate that renewDeadlne < leaseDuration
+	if renewDeadline >= leaseDuration {
+		setupLog.Info("Invalid configuration: LEADER_ELECTION_RENEW_DEADLINE must be less that LEADER_ELECTION_LEASE_DURATION; using default values")
```
Comment (maskarb) [inline cmd/main.go:186]: might be worth adding to these error messages what the values are being overridden to

Comment (djnakabaale) [reply cmd/main.go]: good call 👍🏾, especially for future troubleshooting efforts

**Review by maskarb — APPROVED** [review]

**Pull request merged**